### DISCUSSION TITLE
What do we need to get iterator implementation in `IDictionary`?

### DISCUSSION THREAD

**Discussion by scarletquasar** [dotnet/runtime#91534] [General] [answered]
What is blocking the following code from being a valid iteration implementation?

```cs
public Dictionary<string, object> GetHeaders() 
{
	if (_headers == null) 
	{
		yield return new Dictionary<string, object>();
		yield break;
	}
	
	foreach(var header in _headers) 
	{
		yield return new KeyValuePair<string, object>(header.Key, header.Value);
	}
	
	yield break;
}
```

Note that `Dictionary<TKey, TValue>` implements `IEnumerable<KeyValuePair<TKey, TValue>>`

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

What is blocking the following code from being a valid iteration implementation?

```cs
public Dictionary<string, object> GetHeaders() 
{
	if (_headers == null) 
	{
		yield return new Dictionary<string, object>();
		yield break;
	}
	
	foreach(var header in _headers) 
	{
		yield return new KeyValuePair<string, object>(header.Key, header.Value);
	}
	
	yield break;
}
```

Note that `Dictionary<TKey, TValue>` implements `IEnumerable<KeyValuePair<TKey, TValue>>`

<table>
  <tr>
    <th align="left">Author:</th>
    <td>scarletquasar</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Collections`, `untriaged`, `needs-area-label`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by huoyaoyuan** [comment]
> Note that `Dictionary<TKey, TValue>` implements `IEnumerable<KeyValuePair<TKey, TValue>>`

You can implicitly convert dictionary to IEnumerbale of KVP, but not in the opposite direction. You need to call `ToDictionary` on the result `IEnumerable<KeyValuePair<TKey, TValue>>` at some position.

**Comment by huoyaoyuan** [comment]
> What do we need to get iterator implementation in `IDictionary`?

More closely to the question, the return type of iterator method should be a type that is *implemented by* `IEnumerable<T>` and iteratable, which includes only `IEnumerable<U>` where T : U, and non-generic `IEnumerable` (`object` is not iteratable).

**Comment by scarletquasar** [comment]
@eiriktsarpalis this is an API proposal, why did you convert that to discussion without asking??

**Reply by DaZombieKiller** [reply, marked as answer]
You haven't proposed any new APIs here. You'll need to open a new issue following the [API proposal template](https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.yml&title=%5BAPI+Proposal%5D%3A+).

**Reply by scarletquasar** [reply]
You're wrong. If you read carefully you can see that I'm proposing the possibility of creating a iterator to compose a Dictionary object. After reading the api proposal template I decided that it does not fit in my suggestion because the questions are limited.

**Reply by eiriktsarpalis** [reply]
> If you read carefully you can see that I'm proposing the possibility of creating a iterator to compose a Dictionary object

This is not evident at all reading your original post. In any case, iterators are a language feature so I would recommend creating an issue in the https://github.com/dotnet/csharplang repo.
